frontend/middle-2/design.md

```diff
@@ -1,17 +1,48 @@
 #Базовые принципы проектирования
 
-* Что такое инкапсуляция? Какова мотивация для улучшения уровня инкапсуляции?
-* Что такое абстрагирование? Какие есть отличия между абстракцией, инкапсуляцией и скрытием информации?
-* Что такое coupling? Как инкапсуляция влияет на это?
-* Что такое cohesion? Является ли эта величина обратной величине coupling?
-* Что такое принцип DRY? Как принцип связан с cohesion?
-* Что такое принцип KISS?
-* Что такое принцип YAGNI? Как он соотносится с принципом KISS?
-* Что такое Separation of concerns?
-* Из-за чего возникает проблема impedance mismatch при обмене данными? Какие есть примеры возникновения проблемы в разных технологиях? Какие там используются решения?
-* Почему преждевременная оптимизация может легко обернуться проблемой, а не решением при проектировании систем?
-
+* Inversion of Control
+  * What is IoC?
+  * What are the advantages of frameworks providing IoC over libraries?
+  * What are the advantages of libraries without IoC over frameworks?
+  * Can a module provide an interface that supports both approaches: with IoC and without IoC?
+* Dependency Injection
+  * What is DI?
+  * What are the alternatives to DI? What are the advantages and disadvantages of each alternative?
+  * Can DI be implemented without IoC?
+  * How does DI help to achieve the Separation of Concerns?
+  * How are DI and Dependency Inversion Principle from SOLID related to each other?
+* What is Impedance Mismatch? What are the examples of Impedance Mismatch in different technologies? What are the most popular patterns of solutions?
+* What problems can be caused by premature optimization? How to solve troubles with performance while avoiding premature optimization?
+* What are the benefits of providing an API with different levels of abstraction?
+* Why do callbacks break composability?
+* Large Scale JS
+  * What is the main idea of the proposed approach?
+  * What patterns are used in this approach?
+  * What is the role of events in this approach? Can the similar approach be implemented only with synchronous function invocations?
+  * How does it help to achieve fault tolerance?
+  * How can modules depend on each other?
+* API of modules
+  * What are the characteristics of a good module API?
+  * How can we make an API hard to misuse? What are the benefits of this characteristic?
+  * What are consistency and conceptual integrity properties of API? Why are the properties valuable?
+  * How writing use cases before writing code can help to achieve cleaner API?
+  * What is the difference between extensibility by the module authors and module clients?
+  * How can you explain the phrase "asymmetry of function should be reflected by asymmetry of form"?
+  * What are examples of the naming where close ideas looks similar and different concepts differ greatly?
+  * What are edge cases in the context of designing an API? Why implicit edge case handling is more preferred than explicit?
```

Comment:
What are the edge cases... Why does the implicit edge case handling is preferred to explicit?